process/fuser.rs

```diff
@@ -0,0 +1,1345 @@
+                if pw_entry.is_null() {
+                    "unknownr"
+                } else {
+                    match CStr::from_ptr((*pw_entry).pw_name).to_str() {
+                        Ok(name) => name,
+                        Err(_) => "invalid_string",
+                    }
+                }
+            };
+            eprint!("({})", owner);
+        }
+        io::stderr().flush()?;
+    }
+
+    eprint!("\n");
+    Ok(())
+}
+
+/// Scans the `/proc` directory for process information and checks various access types.
+///
+/// # Arguments
+///
+/// * `need_check_map` - A boolean indicating whether to check the process map files.
+/// * `names` - A mutable reference to a `Names` object for storing matched processes.
+/// * `inode_list` - A reference to an `InodeList` for updating inode information.
+/// * `device_list` - A reference to a `DeviceList` for updating device information.
+/// * `unix_socket_list` - A reference to a `UnixSocketList` for checking Unix sockets.
+/// * `net_dev` - A `u64` representing the network device identifier.
+///
+/// # Errors
+///
+/// Returns an error if reading directory entries, accessing process stats, or checking access types fails.
+///
+/// # Returns
+///
+/// Returns `Ok(())` on success.
+async fn scan_procs(
+    need_check_map: bool,
+    names: &mut Names,
+    inode_list: &InodeList,
+    device_list: &DeviceList,
+    unix_socket_list: &UnixSocketList,
+    net_dev: u64,
+) -> Result<(), io::Error> {
+    let my_pid = std::process::id() as i32;
+    let mut read_dir = tokio::fs::read_dir(PROC_PATH).await?;
+
+    while let Some(entry) = read_dir.next_entry().await? {
+        let filename = entry
+            .file_name()
+            .into_string()
+            .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "Invalid file name"))?;
+
+        if let Ok(pid) = filename.parse::<i32>() {
+            // Skip the current process
+            if pid == my_pid {
+                continue;
+            }
+
+            let cwd_stat = match get_pid_stat(pid, "/cwd") {
+                Ok(stat) => stat,
+                Err(_) => continue,
+            };
+
+            let exe_stat = match get_pid_stat(pid, "/exe") {
+                Ok(stat) => stat,
+                Err(_) => continue,
+            };
+            let root_stat = match get_pid_stat(pid, "/root") {
+                Ok(stat) => stat,
+                Err(_) => continue,
+            };
+
+            let st = timeout(&entry.path().to_string_lossy(), 5)?;
+            let uid = st.st_uid;
+
+            check_root_access(names, pid, uid, &root_stat, device_list, inode_list)?;
+            check_cwd_access(names, pid, uid, &cwd_stat, device_list, inode_list)?;
+            check_exe_access(names, pid, uid, &exe_stat, device_list, inode_list)?;
+
+            if need_check_map {
+                check_map(names, pid, "maps", device_list, uid, Access::Mmap)?;
+            }
+            check_dir(
```

Comment:
All these `check_....` functions should be asynchronous, because then there is no point in making scan_procs asynchronous, `names` will have to be wrapped in a mutex, as far as I can see.